vowpalwabbit/parse_args.cc

```diff
@@ -1897,8 +1762,12 @@ void finish(vw& all, bool delete_all)
     finalize_regressor_exception_thrown = true;
   }
 
-  if (delete_all)
+  // If this object is managed with a unique pointer then we should not delete.
+  // This can happen if a user calls the wrong API or an exception is thrown and the user is using the new api.
+  if (delete_all && !all.is_managed_by_unique)
```

Comment:
Need to think it over, but do we actually need is_managed_by_unique? If the user is using unique_ptr impl then they will call into the correct finish which we control fully by sending delete_all as false and if they aren't is_managed_by_unique will be set to false BUT this leaves room for someone distracted to grab the unique_ptr, get the raw pointer (forget to set is_managed_by_unique to false) and call into this function. Not exactly sure if we could do anything about it but just thinking out loud.

Reply:
Yeah, this situation is kind of annoying. But I think it's just too easy to get it wrong without a check like this.

Consider:
Someone gets their unique_ptr, gets the raw pointer and goes about using it for all of the different APIs. Because, well, the APIs take raw pointers. Then they call finish just like they have always done, and (without this check) it frees the vw object. Then the unique_ptr goes out of scope and delete is called again and we have a double free.

Reply:
But in your scenario the user only gets protected if they are using your API that is the old api wrapping the new api (the one that sets is_managed_by_unique to false) and if they use the new api that returns unique_ptr but later do get() and forget about calling the correct finish it ends up being the same situation (no protection i think?).

Reply:
If they use the old then they must call `finish(vw*)`, there are no other options. They don't have a unique_ptr to call the other variant. 

> if they use the new api that returns unique_ptr but later do get() and forget about calling the correct finish it ends up being the same situation (no protection i think?).

This is the situation I am trying to guard against. They should be protected against a double free here.

Because if `is_managed_by_unique` is true, then they cannot get into the block where the object is deleted.

Reply:
whoops, you are correct since it's the default value 👍